### PR TITLE
tyf/add ROPE

### PR DESCRIPTION
**Pull request by Bonbon-Tang**
## Description
camb支持interleaved和非interleaved所有情况的rotaryembedding计算，目前只支持input是contiguous，contiguous输入下性能达设备最优

## Checklist
**Before PR**:

- [ ] I have read and followed the workflow indicated in the [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: impl/camb/functions_ext/rotaryEmbedding.cpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    DIOPI_CALL(autoCastTensorType(ctx, pTensors, supportedDtypes));
+
+    DiopiTensor outTmpTr = outputTensor;
+    if (outputTensor.dtype() != inputTensor.dtype()) {
+        outTmpTr = requiresTensor(ctx, outputTensor.shape(), outputTensor.stride(), calType);
+    }
+
+    // split input into two part, via offset and stride without new storage;
+    // interleaved == true: abcdefgh -> abcd,efgh in last dimension;
+    // interleaved == false: abcdefgh ->aceg,bdfh in last dimension;
```
Comment (POI-WX) [inline impl/camb/functions_ext/rotaryEmbedding.cpp:48]: 你这里的注释是不是写反了，跟你具体实现逻辑不一致。interleaved是交叉意思。

File: impl/camb/functions_ext/rotaryEmbedding.cpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    void* input2Ptr = static_cast<void*>((static_cast<float*>(inputTensor.data())) + storageOffset);
+    void* output2Ptr = static_cast<void*>((static_cast<float*>(outTmpTr.data())) + storageOffset);
+
+    // set Tensors' decriptor
+    CnnlTensorDesc inputDesc;
+    CnnlTensorDesc cosDesc;
+    CnnlTensorDesc sinDesc;
+    CnnlTensorDesc outTmpDesc;
+
+    if (shape.size() == 1) {
```
Comment (POI-WX) [inline impl/camb/functions_ext/rotaryEmbedding.cpp:73]: 统一描述，b:batch_size, s:seqlen, n:head_num, d:head_dim。一般大家不使用head_size。BSND是最常见的layout。

File: diopi_test/python/configs/diopi_configs.py [hunk 3]
```diff
@@ -8192,11 +8192,11 @@
                 },
                 {
                     "ins": ['cos'],
-                    "shape": ((6,), (32,), (2, 64), (3, 1, 128), (1, 32), (3, 5, 6), (125, 1, 256), (125, 1, 256), (64, 1, 16), (100, 1, 32)),
```
Comment (POI-WX) [inline diopi_test/python/configs/diopi_configs.py:8195]: 这个测例将5改成1，是因为camb不支持head_num为5这种奇数吗？如果camb不支持可以加上check并跳过，实际不会遇到这种测例。

File: impl/camb/functions_ext/rotaryEmbedding.cpp (hunk 2 above)
Comment (POI-WX) [reply impl/camb/functions_ext/rotaryEmbedding.cpp:73]: 这里可以不必拘泥于camb的描述，head一般不说size，这样方便对齐模型层的术语。

File: diopi_test/python/configs/diopi_configs.py (hunk 3 above)
Comment (Bonbon-Tang) [reply diopi_test/python/configs/diopi_configs.py:8195]: 这个测例一方面是不支持，一方面确实不可能出现不同head不同embedding

**Review by POI-WX — APPROVED** [review]

File: impl/camb/functions_ext/rotaryEmbedding.cpp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    } else if (shape.size() == 4) {
+        inputDesc.set(inputTensor.dtype(), shape, stride, CNNL_LAYOUT_ARRAY);
+        outTmpDesc.set(inputTensor.dtype(), shape, stride, CNNL_LAYOUT_ARRAY);
+    } else {
+        DIOPI_CHECK(true, "camb currently only support 1<= input.dim() <= 4");
+    }
+
+    if (cosTensor.dim() == 1) {
+        std::vector<int64_t> cosShape = {1, 1, cosTensor.shape()[0]};
+        std::vector<int64_t> cosStride = calContiguousStride(cosShape);
```
Comment (yangbofun) [inline impl/camb/functions_ext/rotaryEmbedding.cpp:107]: 这里if else中共同的部分可以放到 if else的后面。

**Review by yangbofun — APPROVED** [review]

**Pull request merged**